### PR TITLE
NMS-14596: Import customer edges and links into a requisition

### PR DESCRIPTION
**Pull request by amay94**
JIRA Issue: https://issues.opennms.org/browse/NMS-14596

### CODE REVIEW

File: plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java [hunk 1]
```diff
@@ -81,4 +226,7 @@ public void setEnterpriseId(final UUID enterpriseId) {
         }
     }
 
+    private InetAddress getInetAddress(String ip) {
```
Comment (amay94) [inline plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java:229]: I'll switch to using the Util method once the other PR is merged

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 2]
```diff
@@ -38,12 +40,45 @@ public class Edge {
...
+    public final String lteRegion;
+    public final String logicalId;
+    public final String modelNumber;
+    public final String name;
+    public final Boolean operatorAlertsEnabled;
+    public final String selfMacAddress;
+    public final Integer siteId;
+    public final String softwareVersion;
 
     private Edge(final Builder builder) {
```
Comment (amay94) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:60]: All of the `requireNonNull` here are marked as required in their API

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 3]
```diff
@@ -38,12 +40,45 @@ public class Edge {
...
+        this.deviceId = Objects.requireNonNull(builder.deviceId);
+        this.dnsName = Objects.requireNonNull(builder.dnsName);
+        this.lteRegion = Objects.requireNonNull(builder.lteRegion);
+        this.logicalId = Objects.requireNonNull(builder.logicalId);
+        this.modelNumber = Objects.requireNonNull(builder.modelNumber);
+        this.name = Objects.requireNonNull(builder.name);
+        this.operatorAlertsEnabled = Objects.requireNonNull(builder.operatorAlertsEnabled);
+        this.selfMacAddress = Objects.requireNonNull(builder.selfMacAddress);
+        this.siteId = Objects.requireNonNull(builder.siteId);
+        this.softwareVersion = Objects.requireNonNull(builder.softwareVersion);
```
Comment (amay94) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:80]: Nothing in the `EdgeObject` is marked as required, so I'm not sure which of these we can count on to be populated.

Comment (aramosvizcarra) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:80]: I think we just need to play by ear and once we are able to test against the dev environment we'll figure that out. I did everything required but from the API documentation doesn't seem that many fields are.

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 4]
```diff
@@ -85,6 +154,86 @@ public Builder withHub(final boolean hub) {
...
+        public Builder withSiteId(Integer siteId) {
+            this.siteId = siteId;
+            return this;
+        }
+
+        public Builder withSoftwareVersion(String softwareVersion) {
+            this.softwareVersion = softwareVersion;
+            return this;
+        }
+
```
Comment (aramosvizcarra) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:236]: I think we need a toString() method also

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+        public Link build() {
+            return new Link(this);
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+}
```
Comment (aramosvizcarra) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java:258]: Same here toString()

**Review by aramosvizcarra — CHANGES_REQUESTED** [review]
LGTM, just a couple of changes

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java (hunk 4 above)
Comment (amay94) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:236]: Do we need one in the builder? The outer class has one on line 85-110

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java (hunk 5 above)
Comment (amay94) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java:258]: See lines 87-110

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java (hunk 4 above)
Comment (aramosvizcarra) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:236]: Never mind I didn't see at the top.

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java (hunk 5 above)
Comment (aramosvizcarra) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java:258]: yes sorry I'm used to find it at the end. Didn't see it at the top :)

**Review by aramosvizcarra — DISMISSED** [review]
LGTM

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 6]
```diff
@@ -38,12 +40,45 @@ public class Edge {
     public final String operator;
     public final String site;
     public final boolean hub;
+    public final List<Link> links;
+    public final Boolean alertsEnabled;
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:44]:
```suggestion
    public final boolean alertsEnabled;
```

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 7]
```diff
@@ -38,12 +40,45 @@ public class Edge {
...
+    public final String customInfo;
+    public final String description;
+    public final String deviceFamily;
+    public final String deviceId;
+    public final String dnsName;
+    public final String lteRegion;
+    public final String logicalId;
+    public final String modelNumber;
+    public final String name;
+    public final Boolean operatorAlertsEnabled;
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:55]:
```suggestion
    public final boolean operatorAlertsEnabled;
```

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 8]
```diff
@@ -38,12 +40,45 @@ public class Edge {
...
+    public final String selfMacAddress;
+    public final Integer siteId;
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:57]:
```suggestion
    public final int siteId;
```

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java (hunk 3 above)
Comment (fooker) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:80]: I think having this as required is fine for now.

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 9]
```diff
@@ -53,6 +88,24 @@ public String toString() {
                           .add("operator", this.operator)
                           .add("site", this.site)
                           .add("hub", this.hub)
+                          .add("links", this.links.stream()
+                                                        .map(l -> l.toString())
+                                                        .collect(Collectors.joining(",")))
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:93]:
```suggestion
                          .add("links", this.links)
```

Should do the same.

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 10]
```diff
@@ -61,6 +114,22 @@ public static class Builder {
         private String operator;
         private String site;
         private boolean hub;
+        private List<Link> links;
+        private Boolean alertsEnabled;
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:118]:
```suggestion
        private boolean alertsEnabled;
```

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 11]
```diff
@@ -61,6 +114,22 @@ public static class Builder {
...
+        private String customInfo;
+        private String description;
+        private String deviceFamily;
+        private String deviceId;
+        private String dnsName;
+        private String lteRegion;
+        private String logicalId;
+        private String modelNumber;
+        private String name;
+        private Boolean operatorAlertsEnabled;
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:129]:
```suggestion
        private boolean operatorAlertsEnabled;
```

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 12]
```diff
@@ -61,6 +114,22 @@ public static class Builder {
...
+        private String selfMacAddress;
+        private Integer siteId;
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:131]:
```suggestion
        private int siteId;
```

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+
+import com.google.common.base.MoreObjects;
+
+import java.time.OffsetDateTime;
+import java.util.Objects;
+import java.util.UUID;
+
+public class Link {
+    public final UUID enterpriseId;
+    public final Integer id;
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Link.java:39]: Same here. As soon as we checking for non-null, this should be the native types. Same goes for builders.

File: plugin/src/main/java/org/opennms/velocloud/requisition/AbstractRequisitionProvider.java [hunk 14]
```diff
@@ -48,6 +48,10 @@
 
     private final VelocloudApiClientProvider clientProvider;
 
+    public static final String VELOCLOUD_METADATA_CONTEXT = "velocloud";
+
+    public static final String VELOCLOUD_LINK_SERVICE_NAME = "velocloud_link";
```
Comment (fooker) [inline plugin/src/main/java/org/opennms/velocloud/requisition/AbstractRequisitionProvider.java:53]:
```suggestion
    public static final String VELOCLOUD_LINK_SERVICE_NAME = "VelocloudLink";
```

to follow common service naming.

File: plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java (hunk 1 above)
Comment (fooker) [reply plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java:229]: What PR is this talking about?

**Review by fooker — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java (hunk 1 above)
Comment (amay94) [reply plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java:229]: [This](https://github.com/OpenNMS/opennms-velocloud-plugin/pull/11/files#diff-cb63cf1591b1dd5f454cda6acc2c87bf84544753f0d66637f26d6ff195c672fdR34) method from Alberto's PR.

**Review by aramosvizcarra — DISMISSED** [review]
LGTM!

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java [hunk 15]
```diff
@@ -126,26 +124,26 @@ public static class Builder {
...
-        private Integer siteId;
+        private int siteId;
         private String softwareVersion;
 
         private Builder() {
         }
 
         public Builder withEnterpriseId(final UUID enterpriseId) {
-            this.enterpriseId = enterpriseId;
+            this.enterpriseId = Objects.requireNonNull(enterpriseId);
```
Comment (fooker) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:136]: This is not good style. If this setter is never called on the builder, the `enterpriseId` will still be `null`. While building, `null` is fine. Therefore the check should be moved back to the model.

Comment (fooker) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/Edge.java:136]: If we do not know about a default for the native types, using the object version in the builder is fine, too. But it should be native in the model.

**Review by fooker — CHANGES_REQUESTED** [review]

**Review by aramosvizcarra — DISMISSED** [review]
Looks good to me

**Review by fooker — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java [hunk 16]
```diff
@@ -61,6 +75,152 @@ protected Requisition handleRequest(final Request request, final VelocloudApiCli
         final var requisition = ImmutableRequisition.newBuilder()
                                                     .setForeignSource(request.getForeignSource());
 
+        try {
+            client.getEdges(request.getEnterpriseId().toString()).stream().forEach(edge -> {
```
[the remark this answers is not in the document]
Comment (amay94) [reply plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java:79]: Should the request throw an exception in the constructor or in getEnterpriseId()? I would think the constructor, but then we would need to catch it in handleRequest(), unless we throw an unchecked exception.

[the remark this answers is not in the document]
Comment (amay94) [reply plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java]: I pushed a version that throws the exception in getEnterpriseId() so we can handle it when handling the request. I don't love it, but IMO it's better than the alternative

File: plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java [hunk 17]
```diff
@@ -69,19 +229,22 @@ public static class Request extends AbstractRequisitionProvider.Request {
...
+            this(connection, connection.getEnterpriseId());
         }
 
-        private UUID enterpriseId;
+        public Request(final Connection connection, Optional<UUID> enterpriseId) {
+            super(TYPE, connection);
+            this.enterpriseId = enterpriseId;
+        }
+
+        private Optional<UUID> enterpriseId;
```
Comment (fooker) [inline plugin/src/main/java/org/opennms/velocloud/requisition/CustomerRequisitionProvider.java:240]: `Optional` is only meant to be a return value (or an intermediate variable). Reason for this is: with `Optional` in a field, there are two kinds of `null` value now - which makes things even worse.

File: client/v1/src/main/java/org/opennms/velocloud/client/v1/VelocloudApiClientV1.java [hunk 18]
```diff
@@ -146,13 +148,54 @@ public List<Edge> getEdges(final String enterpriseId) throws VelocloudApiExcepti
...
             return edges.stream()
-                    .map(e -> Edge.builder()
-                            .withEnterpriseId(enterpriseId)
-                            .withSite(e.getSite().getName())
-                            .withOperator(e.getConfiguration().getOperator().getName())
-                            .withHub(e.isIsHub())
-                            .build())
-                    .collect(Collectors.toList());
+                        .map(e -> Edge.builder()
+                                      .withEnterpriseId(enterpriseId)
```
Comment (fooker) [inline client/v1/src/main/java/org/opennms/velocloud/client/v1/VelocloudApiClientV1.java:152]: This is somewhat bogus. A tenant connection always has some kind of intrinsic enterprise ID. But blindly setting it from the parameter is wrong. TBH, the whole way we handle the enterprise ID in tenant connections is wrong.

**Review by fooker — CHANGES_REQUESTED** [review]

File: client/v1/src/main/java/org/opennms/velocloud/client/v1/VelocloudApiClientV1.java (hunk 18 above)
Comment (amay94) [reply client/v1/src/main/java/org/opennms/velocloud/client/v1/VelocloudApiClientV1.java:152]: My understanding is that when the customer/tenant credentials are stored, the enterpriseId of their system is saved with them. When we get a request with an alias to those customer's credentials, we get that enterpriseId from SCV to use. Individual requests should never be able to specify the enterpriseId, just the alias of the credentials to use.

**Review by fooker — APPROVED** [review]

**Pull request merged**